src/runtime/cuda.cpp

```diff
@@ -589,39 +621,16 @@ WEAK int halide_cuda_copy_to_device(void *user_context, buffer_t* buf) {
     uint64_t t_before = halide_current_time_ns(user_context);
     #endif
 
-    halide_assert(user_context, buf->host && buf->dev);
+    halide_assert(user_context, buf->host && buf->device);
     halide_assert(user_context, validate_device_pointer(user_context, buf));
 
     device_copy c = make_host_to_device_copy(buf);
 
-    // TODO: Is this 32-bit or 64-bit? Leaving signed for now
-    // in case negative strides.
-    for (int w = 0; w < (int)c.extent[3]; w++) {
-        for (int z = 0; z < (int)c.extent[2]; z++) {
-            for (int y = 0; y < (int)c.extent[1]; y++) {
-                for (int x = 0; x < (int)c.extent[0]; x++) {
-                    uint64_t off = (x * c.stride_bytes[0] +
-                                    y * c.stride_bytes[1] +
-                                    z * c.stride_bytes[2] +
-                                    w * c.stride_bytes[3]);
-                    void *src = (void *)(c.src + off);
-                    CUdeviceptr dst = (CUdeviceptr)(c.dst + off);
-                    uint64_t size = c.chunk_size;
-                    debug(user_context) << "    cuMemcpyHtoD "
-                                        << "(" << x << ", " << y << ", " << z << ", " << w << "), "
-                                        << src << " -> " << (void *)dst << ", " << size << " bytes\n";
-                    CUresult err = cuMemcpyHtoD(dst, src, size);
-                    if (err != CUDA_SUCCESS) {
-                        error(user_context) << "CUDA: cuMemcpyHtoD failed: "
-                                            << get_error_name(err);
-                        return err;
-                    }
-                }
-            }
-        }
+    int err = do_multidimensional_copy(user_context, c, c.dst, c.src, buf->dimensions, false);
```

Comment:
Shouldn't this be c.dimensions? (Or just grabbed inside the function.) I think the dimensionality of a device_copy object could be less than the dimensionality of the buffer. Maybe this is safe because device_copy replaces the optimized extents with 1?